File: skills/adkit/scripts/src/ideas/parse.test.ts
import { describe, expect, it } from "vitest";
import { extractNegatives, readThemeGroups } from "./parse.js";

describe("readThemeGroups", () => {
  it("makes one ad group per Keyword Theme in file order, name stripped of role note", () => {
    const md = `
## Go To Market

### Keyword Themes

#### Widget Software — category core

- widget software
- widget platform

#### Free / freemium intent

- free widget tool
`;
    expect(readThemeGroups(md, 20)).toEqual([
      ["Widget Software", ["widget software", "widget platform"]],
      ["Free / freemium intent", ["free widget tool"]],
    ]);
  });

  it("strips offer suffix and markdown from theme keywords", () => {
    const md = `
## Go To Market

### Keyword Themes

#### Agency

- hire widget agency — offer: 15-minute walkthrough
- *widget pricing* now (12k, HIGH, $2–$8)
`;
    expect(readThemeGroups(md, 20)).toEqual([["Agency", ["hire widget agency", "widget pricing now"]]]);
  });

  it("excludes [spend-trap] themes regardless of marker position or case", () => {
    const before = `
## Go To Market

### Keyword Themes

#### Core

- widget software

#### [spend-trap] Generic Scheduling — keep-but-don't-lead

- scheduling tool
`;
    expect(readThemeGroups(before, 20)).toEqual([["Core", ["widget software"]]]);

    const after = `
## Go To Market

### Keyword Themes

#### Generic Scheduling [Spend-Trap] — keep-but-don't-lead

- scheduling tool

#### Core

- widget software
`;
    expect(readThemeGroups(after, 20)).toEqual([["Core", ["widget software"]]]);
  });

  it("dedups a keyword across themes (first-seen wins — no cannibalization)", () => {
    const md = `
## Go To Market

### Keyword Themes

#### Theme A

- shared widget
- only a

#### Theme B

- shared widget
- only b
`;
    expect(readThemeGroups(md, 20)).toEqual([
      ["Theme A", ["shared widget", "only a"]],
      ["Theme B", ["only b"]],
    ]);
  });

  it("caps keywords per theme", () => {
    const md = `
## Go To Market

### Keyword Themes

#### Theme

- one
- two
- three
- four
`;
    expect(readThemeGroups(md, 2)).toEqual([["Theme", ["one", "two"]]]);
  });

  it("returns empty when there is no ### Keyword Themes section (create then requires a gtm re-run)", () => {
    const onlyTiers = `
## Go To Market

### Keywords

#### Commercial

- buy widgets
`;
    expect(readThemeGroups(onlyTiers, 20)).toEqual([]);
    expect(readThemeGroups("## Something Else\n\n- foo\n", 20)).toEqual([]);
  });
});

describe("extractNegatives", () => {
  it("parses the section and strips the reason", () => {
    const md = `
## Go To Market

### Keywords

#### Commercial

- buy widgets

#### Negative Keywords

- jobs — reason: job seekers
- *free* download
- near me
`;
    expect(extractNegatives(md)).toEqual([
      { text: "jobs", matchType: "PHRASE" },
      { text: "free download", matchType: "PHRASE" },
      { text: "near me", matchType: "PHRASE" },
    ]);
  });

  it("returns empty when absent", () => {
    const md = "## Go To Market\n\n### Keywords\n\n#### Commercial\n\n- buy widgets\n";
    expect(extractNegatives(md)).toEqual([]);
  });
});

File: skills/adkit/scripts/ads_skill/ideas/parse_test.py
from __future__ import annotations

from ads_skill.ideas.parse import _extract_negatives, _read_theme_groups


def test_read_theme_groups_one_ad_group_per_tier_in_order() -> None:
    md = """
## Go To Market

### Keywords

#### Informational

- what is a widget

#### Commercial

- buy widgets online
- compare widget prices

#### Transactional

- widget checkout
"""
    themes = _read_theme_groups(md, 20)
    assert themes == [
        ("Informational", ["what is a widget"]),
        ("Commercial", ["buy widgets online", "compare widget prices"]),
        ("Transactional", ["widget checkout"]),
    ]


def test_read_theme_groups_strips_offer_suffix_and_markdown() -> None:
    md = """
## Go To Market

### Keywords

#### Commercial

- hire widget agency — offer: 15-minute walkthrough
- *widget pricing* now
"""
    assert _read_theme_groups(md, 20) == [
        ("Commercial", ["hire widget agency", "widget pricing now"]),
    ]


def test_read_theme_groups_reads_tiers_verbatim() -> None:
    # The reader makes NO grouping decision — it trusts ads:gtm to put each
    # keyword in exactly one tier and reads each tier's bullets as authored.
    md = """
## Go To Market

### Keywords

#### Commercial

- widget pricing
- compare widgets

#### Transactional

- buy widgets
"""
    assert _read_theme_groups(md, 20) == [
        ("Commercial", ["widget pricing", "compare widgets"]),
        ("Transactional", ["buy widgets"]),
    ]


def test_read_theme_groups_caps_keywords_per_theme() -> None:
    md = """
## Go To Market

### Keywords

#### Commercial

- one
- two
- three
- four
"""
    assert _read_theme_groups(md, 2) == [("Commercial", ["one", "two"])]


def test_read_theme_groups_empty_when_no_gtm_section() -> None:
    assert _read_theme_groups("## Something Else\n\n- foo\n", 20) == []


def test_extract_negatives_parses_section_strips_reason() -> None:
    md = """
## Go To Market

### Keywords

#### Commercial

- buy widgets

#### Negative Keywords

- jobs — reason: job seekers
- *free* download
- near me
"""
    assert _extract_negatives(md) == [
        {"text": "jobs", "matchType": "PHRASE"},
        {"text": "free download", "matchType": "PHRASE"},
        {"text": "near me", "matchType": "PHRASE"},
    ]


def test_extract_negatives_empty_when_absent() -> None:
    md = "## Go To Market\n\n### Keywords\n\n#### Commercial\n\n- buy widgets\n"
    assert _extract_negatives(md) == []
